### ISSUE TITLE
Official ONNX Export Script for Dynamic Batching

### ISSUE THREAD

**Issue by Mahaaveer** [roboflow/rf-detr#376]
### Search before asking

- [x] I have searched the RF-DETR [issues](https://github.com/roboflow/rf-detr/issues) and found no similar feature requests.

### Description

Dear RF-DETR Team,

First off, thank you for creating and maintaining this fantastic project. The RF-DETR Nano model is incredibly efficient and i am looking to use it in a multi-stream video analytics pipeline.

**The Goal:**

The goal is to convert the RF-DETR Nano model to a TensorRT engine that supports dynamic batch sizes. This is critical for maximizing GPU throughput in a real-time application that processes frames from multiple camera streams simultaneously.

**The Problem:**

The current default export method (model.export()) produces a static ONNX file with a hardcoded batch size of 1. When i tried to convert this to a TensorRT engine with dynamic shapes, trtexec fails as expected.

I spent considerable time trying to create a custom export script to produce a valid, dynamic ONNX file, but have run into a series of issues that suggest the model's architecture contains patterns that are incompatible with the standard torch.onnx.export tracing process when dynamic axes are used.

**Summary of Attempts and Findings:**

**Initial Attempt (Dynamic Axes)**: I created a custom export script using torch.onnx.export with dynamic_axes set for the batch dimension. This failed during trtexec conversion with a reshape would change volume error, indicating an internal Reshape operation was using a hardcoded shape based on the trace-time batch size of 1.

**Input Type Mismatch**: We then discovered the model's forward pass expects a custom NestedTensor object, not a raw tensor. The tracer was failing because it was providing the wrong input type.

**Wrapper Approach**: I implemented the standard "ONNX Export Wrapper" pattern. This correctly constructs the NestedTensor from raw tensor inputs, satisfying the model's API. This successfully resolved the input type errors.

**Unsupported Operator**: The wrapper allowed the tracer to go deeper, where it found an unsupported operator aten::_upsample_bicubic2d_aa. I solved this by registering a custom symbolic function to map it to the standard ONNX Resize operator.

**Final trtexec Error**: Even with a seemingly perfect ONNX file, trtexec ultimately fails during parsing with an INVALID_NODE error on a ConvTranspose layer, stating that the number of input channels cannot be dynamic. We have also seen Concat layer errors where a cls_token with a static batch size of 1 is concatenated with feature maps that have a dynamic batch size.

**Conclusion:**

These findings strongly suggest that the RF-DETR architecture contains several operations (like ConvTranspose with dynamic channels or hardcoded constants like cls_token) that make it fundamentally incompatible with a straightforward dynamic batch export. 

**Feature Request:**

Would it be possible for the development team to provide an official, supported Python script for exporting the RF-DETR models to ONNX with dynamic batch support that goes well with trtexec?

Having an official script would be a massive benefit to the community looking to deploy these models in high-performance, multi-stream environments. It would ensure that the known architectural hurdles are correctly handled during the export process.

Thank you for your consideration and for all your work on this project. 

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by knefenk** [comment]
> **Unsupported Operator**: The wrapper allowed the tracer to go deeper, where it found an unsupported operator aten::_upsample_bicubic2d_aa. I solved this by registering a custom symbolic function to map it to the standard ONNX Resize operator.

Just adding on top of that, I have tried exporting to onnx but have similar issue, solved it by changing it to "bilinear" as below:
```
        patch_pos_embed = nn.functional.interpolate(
            patch_pos_embed.to(dtype=torch.float32),
            size=(torch_int(height), torch_int(width)),  # Explicit size instead of scale_factor
            mode="bilinear",
            align_corners=False,
            # antialias=True,
        ).to(dtype=target_dtype)
```
in the following directory: .venv\Lib\site-packages\rfdetr\models\backbone\dinov2_with_windowed_attn.py  and .venv\Lib\site-packages\rfdetr\models\backbone\dinov2.py

not sure if that helps, just want to provide more information on this matter

**Comment by isaacrob-roboflow** [comment]
Again, that layer shouldn't even be in the exported graph. It's only there to resize the positional embedding for different images resolutions, which is fixed at export time. Are you exporting the official way and seeing this layer present?

**Comment by knefenk** [comment]
Thanks for the clarification. I tried to recreate the export using what I believe is the official method, do please let me know if I’ve misunderstood anything, I’m sharing only the relevant part of the code here:
```
        print(f" Loading model from {checkpoint}")
        model = RFDETRNano(pretrain_weights=checkpoint)

        # Run export
        print(f"Exporting to ONNX...")
        model.export(
            output_dir=MODEL_FOLDER,
            simplify=SIMPLIFY,
            shape=EXPORT_SHAPE,
            verbose=True
        )
```
However, I still ran into the following error:
`Failed to export xxx : Exporting the operator 'aten::_upsample_bicubic2d_aa' to ONNX opset version 17 is not supported
`
Just wanted to check if this matches what you mean by the official export path, or if I may be missing a step.

**Comment by knefenk** [comment]
Just a quick update, the export works fine after commenting the "antialias=True," or changing it to False.
#
> ```
>         patch_pos_embed = nn.functional.interpolate(
>             patch_pos_embed.to(dtype=torch.float32),
>             size=(torch_int(height), torch_int(width)),  # Explicit size instead of scale_factor
>             mode="bicubic",
>             align_corners=False,
>             # antialias=True,
>         ).to(dtype=target_dtype)
> ```

Hope this helps

**Comment by igm503** [comment]
Bump

IMO this is a pretty important neglected issue. RF-DETR is a great model, but difficulties exporting with dynamic batch support meaningfully reduce its usefulness. 

I'd be happy to take a stab at contributing to fix this, but I'm hesitant because it seems that past PRs related to this issue have gone stale. Is this something the team has the bandwidth to review if a fix is provided?

**Comment by Isabelam00** [comment]
I would love for this thread to end with a solution; I’m also dealing with the same problem.

**Comment by FrancescoCappio** [comment]
I was able to export the segmentation model with this script: 
```python
#!/usr/bin/env python3
import torch
import torch.nn as nn
import torch.nn.functional as F
from rfdetr import RFDETRSegNano

from rfdetr.models.backbone import projector


# A LayerNorm fix is needed for dynamic batching support
# We apply the fix before creating the model
def fixed_forward(self, x):
    x = x.permute(0, 2, 3, 1)
    x = F.layer_norm(x, self.normalized_shape, self.weight, self.bias, self.eps)
    x = x.permute(0, 3, 1, 2)
    return x


projector.LayerNorm.forward = fixed_forward

RESOLUTION = 672


class RFDETRSegONNXWrapper(nn.Module):
    def __init__(self, rfdetr_model):
        super().__init__()
        self.model = rfdetr_model.model.model  # Access the underlying PyTorch module
        self.resolution = rfdetr_model.model.resolution

    def forward(self, x):
        # 1. Standard Forward Pass
        # Returns: {'pred_logits': ..., 'pred_boxes': ..., 'pred_masks': ...}
        outputs = self.model(x) # exported model output is a tuple not a dict
        
        logits = outputs[1] #['pred_logits']
        boxes = outputs[0] #['pred_boxes']
        mask_logits = outputs[2] #['pred_masks'] # These are often low-res prototypes

        # 2. Post-processing: Apply Sigmoid to get probabilities
        probs = logits.sigmoid()
        
        # 3. Mask Decoding (Simplest form: Sigmoid + Upsample)
        # Note: If RF-DETR uses a prototype-based head (like Mask-DINO), 
        # the 'pred_masks' are already the combined result but at low resolution.
        masks = mask_logits.sigmoid()
        
        # Bilinear upsample to match input resolution (e.g., 312x312)
        # We use a fixed size to ensure ONNX compatibility
        masks = nn.functional.interpolate(
            masks,
            size=(self.resolution, self.resolution),
            mode='bilinear',
            align_corners=False
        )

        return probs, boxes, masks


# --- Execution Script ---

# 1. Load your trained model
model_path = "rfdetr_segm_output_5e/checkpoint_best_total.pth"
rf_model = RFDETRSegNano(pretrain_weights=model_path, resolution=RESOLUTION) # training time resolution is not stored in the ckpt
rf_model.model.model.eval()  # needed to get the predicted masks from the inner model
rf_model.model.model.export()  # convert layers to their export-able version

# 2. Initialize Wrapper
wrapper = RFDETRSegONNXWrapper(rf_model)
wrapper.eval()

# 3. Define Dummy Input (Matching Nano's default 312 or your custom resolution)
dummy_input = torch.randn(1, 3, rf_model.model.resolution, rf_model.model.resolution).cuda()

# 4. Export
onnx_file = "rf_detr_seg_nano_full.onnx"

torch.onnx.export(
    wrapper,
    dummy_input,
    onnx_file,
    export_params=True,
    opset_version=17,
    keep_initializers_as_inputs=False,
    do_constant_folding=True,
    input_names=['input'],
    output_names=['scores', 'boxes', 'masks'],
    dynamic_axes={'input': {0: 'batch_size'}, 'scores': {0: 'batch_size'}, 'boxes': {0: 'batch_size'}, 'masks': {0: 'batch_size'}},
    verbose=False
)

print(f"Model exported with post-processing to {onnx_file}")
```

I can confirm the output works hosted on a Triton server

**Comment by dresiwela** [comment]
Is there any official update on this or has anyone been able to find a solution? I have struggling with the exact same issue as we need adaptive batching on nvinfer for our production deepstream pipeline.

**Comment by svengoluza** [comment]
I manage to export it to ONNX with fixes recommended in the comments, but when creating the TensorRT engine, I get the warning (i set batch size to 8):

`[03/12/2026-14:23:42] [TRT] [W] Profile kMAX values are not self-consistent. IShuffleLayer /model/transformer/Reshape_5: reshaping failed for tensor: onnx::Reshape_2668_output reshape would change volume 2 to 16 In executeReshape at /_src/runtime/core/shapeMachine.cpp:347 Instruction: RESHAPEinput dims{1 2} reshape dims{8 1 1 2}. In executeContinuation at /_src/runtime/core/shapeMachine.cpp:1101`


And later, when running inference with batched input, I get the error:

`[03/12/2026-14:25:09] [TRT] [E] IExecutionContext::getTensorShape: Error Code 7: Internal Error (IShuffleLayer /model/transformer/Reshape_5: reshaping failed for tensor: onnx::Reshape_2668_output reshape would change volume 2 to 16 In executeReshape at /_src/runtime/core/shapeMachine.cpp:347 Instruction: RESHAPEinput dims{1 2} reshape dims{8 1 1 2}. In executeContinuation at /_src/runtime/core/shapeMachine.cpp:1101)`


I create TensorRT engine using this function:

```
def convert_to_tensorrt(
    onnx_path: Path,
    fp16: bool,
    workspace_gb: int = 4,
    max_batch_size: int = 1,
) -> Path:
    """
    Convert an ONNX model to a TensorRT engine using the Python Builder API.

    Must be run on the same GPU family used for deployment.

    :param onnx_path: Path to the ONNX model
    :param fp16: Enable FP16 precision
    :param workspace_gb: Max workspace memory in GB
    :param max_batch_size: Upper bound for dynamic batch dimension
    :return: Path to the generated .engine file
    """
    onnx_path = Path(onnx_path)
    engine_path = onnx_path.with_suffix(".engine")

    trt_logger = trt.Logger(trt.Logger.WARNING)
    trt.init_libnvinfer_plugins(trt_logger, "")
    builder = trt.Builder(trt_logger)
    config = builder.create_builder_config()
    config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, workspace_gb << 30)

    network = builder.create_network(1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH))
    parser = trt.OnnxParser(network, trt_logger)

    logger.info(f"Parsing ONNX model: {onnx_path}")
    success = parser.parse_from_file(str(onnx_path))
    for i in range(parser.num_errors):
        logger.error(f"  TensorRT parser error: {parser.get_error(i)}")
    if not success:
        raise RuntimeError(f"Failed to parse ONNX model: {onnx_path}. Success: {success}")

    profile = builder.create_optimization_profile()
    for i in range(network.num_inputs):
        input_tensor = network.get_input(i)
        input_shape = input_tensor.shape

        if any(d < 0 for d in input_shape):
            min_shape = (1, 3, 256, 256)
            opt_shape = (1, 3, 720, 1280)
            max_shape = (max_batch_size, 3, 1080, 2048)
            logger.info(f"Setting profile for {input_tensor.name}: min={min_shape}, opt={opt_shape}, max={max_shape}")
            profile.set_shape(input_tensor.name, min_shape, opt_shape, max_shape)

    config.add_optimization_profile(profile)

    if fp16 and builder.platform_has_fast_fp16:
        config.set_flag(trt.BuilderFlag.FP16)
        logger.info("FP16 precision enabled")

    logger.info("Building TensorRT engine (this may take a few minutes)...")
    serialized_engine = builder.build_serialized_network(network, config)
    if serialized_engine is None:
        raise RuntimeError("TensorRT engine build failed")

    with open(engine_path, "wb") as f:
        f.write(serialized_engine)

    logger.info(f"TensorRT engine saved: {engine_path}")
    return engine_path
```

**Comment by isaacrob** [comment]
@igm503 we've brought on @Borda who is now helping to make sure things don't go stale, and the repo is moving forward in a great direction :) if you'd like to contribute, we have much more capacity now

**Comment by raghav-rathi** [comment]
I'd like to take this. State on `develop` @ 6b91a48e, checked today on a 2x RTX 4090 box with TensorRT 11.3 / polygraphy 0.53:

- The ONNX side is done. `RFDETRNano().export(dynamic_batch=True)` and the Seg-Nano equivalent run at batch 1, 2, 3, 4, 5, 6 through ONNX Runtime and every image matches eager PyTorch to ~1e-4 (`dets`/`labels`, and `masks` for seg). The only large diffs I saw were the two-stage top-k near-tie flips on random-noise inputs the CHANGELOG already describes, and they are batch-position independent.
- `format="tensorrt"` refuses `dynamic_batch=True` purely because `TensorRTExporter._compile` passes no optimization profile to polygraphy (`CreateConfig(fp16=...)` only), so the engine is fixed to the traced batch.

Plan for the PR:

1. `TensorRTConfig` gains `opt_batch_size` (fed from the existing `batch_size` export kwarg) and a new `max_batch_size` kwarg on `RFDETR.export()`; with `dynamic_batch=True` the engine is built with `Profile().add("input", min=(1,3,H,W), opt=(opt,3,H,W), max=(max,3,H,W))`, taking H/W from the parsed network so a standalone `build_engine(dynamic.onnx)` works too. Static exports keep the exact call they make today.
2. Flip `supports_dynamic_batch` in the exporter and the registry (the mirror is test-enforced), drop the "compiled without an optimization profile" reason, update `docs/learn/export.md`.
3. `TRTInference` (the benchmark helper) allocates bindings at the profile's max batch and calls `set_input_shape` per call instead of raising `NotImplementedError` on a `-1` dim.
4. Tests: unit (profile shapes handed to polygraphy, `max_batch_size` validation) plus an `e2e_tensorrt` case that builds one dynamic Nano engine and checks batch 1 and 4 against eager. Benchmark table from the 4090 (static vs dynamic engine, batch 1/4/8, fp16) in the PR body.

Shout if you'd rather see a different kwarg shape; otherwise I'll open the PR against `develop` shortly.
